crates/binstalk-downloader/src/gh_api_client/request.rs

```diff
@@ -102,27 +131,133 @@ pub(super) async fn fetch_release_artifacts(
     let status = response.status();
     let headers = response.headers();
 
-    if status == remote::StatusCode::FORBIDDEN
-        && headers
-            .get("x-ratelimit-remaining")
-            .map(|val| val == "0")
-            .unwrap_or(false)
-    {
-        return Ok(FetchReleaseRet::ReachedRateLimit {
-            retry_after: headers.get("x-ratelimit-reset").and_then(|value| {
-                let secs = value.to_str().ok()?.parse().ok()?;
-                Some(Duration::from_secs(secs))
-            }),
-        });
+    if let Some(ret) = check_for_status(status, headers) {
+        return Ok(ret);
     }
 
-    if status == remote::StatusCode::UNAUTHORIZED {
-        return Ok(FetchReleaseRet::Unauthorized);
+    Ok(FetchReleaseRet::Artifacts(response.json().await?))
+}
+
+#[derive(Deserialize)]
+struct GraphQLResponse {
+    data: GraphQLData,
+}
+
+#[derive(Deserialize)]
+struct GraphQLData {
+    repository: Option<GraphQLRepo>,
+}
+
+#[derive(Deserialize)]
+struct GraphQLRepo {
+    release: Option<GraphQLRelease>,
+}
+
+#[derive(Deserialize)]
+struct GraphQLRelease {
+    #[serde(rename = "releaseAssets")]
+    assets: GraphQLReleaseAssets,
+}
+
+#[derive(Deserialize)]
+struct GraphQLReleaseAssets {
+    nodes: Vec<Artifact>,
+    #[serde(rename = "pageInfo")]
+    page_info: GraphQLPageInfo,
+}
+
+#[derive(Deserialize)]
+struct GraphQLPageInfo {
+    #[serde(rename = "endCursor")]
+    end_cursor: CompactString,
+    #[serde(rename = "hasNextPage")]
+    has_next_page: bool,
+}
+
+enum FilterCondition {
+    Init,
+    After(CompactString),
+}
+
+impl fmt::Display for FilterCondition {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            // GitHub imposes a limit of 100 for the value passed to param "first"
+            FilterCondition::Init => f.write_str("first:100"),
+            FilterCondition::After(end_cursor) => write!(f, "after:\\\"{end_cursor}\\\""),
+        }
     }
+}
 
-    if status == remote::StatusCode::NOT_FOUND {
-        return Ok(FetchReleaseRet::ReleaseNotFound);
+async fn fetch_release_artifacts_graphql_api(
+    client: &remote::Client,
+    GhRelease { owner, repo, tag }: &GhRelease,
+    auth_token: &str,
+) -> Result<FetchReleaseRet, GhApiError> {
+    static GRAPHQL_ENDPOINT: OnceLock<Url> = OnceLock::new();
+
+    let graphql_endpoint = GRAPHQL_ENDPOINT.get_or_init(|| {
+        Url::parse("https://api.github.com/graphql").expect("Literal provided must be a valid url")
+    });
+
+    let mut artifacts = Artifacts::default();
+    let mut cond = FilterCondition::Init;
+
+    loop {
+        // NOTE: Don't forget to escape quotation in json value!
+        let graphql_query = format!(
+            r#"
+{{"query":"query {{ repository(owner:\"{owner}\",name:\"{repo}\") {{ release(tagName:\"{tag}\") {{ releaseAssets({cond}) {{ nodes {{ name }} pageInfo {{ endCursor hasNextPage }} }} }} }} }}"}}"#
```

Comment:
Could you put this on several lines so it's easier to read

Reply:
Also can we use serde_json to make the json and write straight graphql instead of escaping quotes

Reply:
> Could you put this on several lines so it's easier to read

I tried and failed since github obviously cannot parse multi-line value.

Reply:
I think that's because of the 'manually make JSON' thing.

Reply:
> I think that's because of the 'manually make JSON' thing.

True, I've switched to `serde_json` and it just works.